## Background

As the scale of genomic sequencing continues to grow, fast and low-memory methods for estimating sequence similarity have become critical for conducting tasks ranging from taxonomic classification to phylogenetic analysis on large-scale datasets [@mash; @kraken2].
Alignment-free methods using k-mers, particularly when used with subsampling methods such as MinHash [@mash] and derivates such as FracMinHash [@sourmash_gather] can quickly compare and classify metagenomic datasets.
However, nucleotide k-mer analyses can suffer from limited sensitivity when comparing highly polymorphic sequences or classifying organisms from groups that are not well represented in reference databases.
Indeed, current classification methods often only categorize a small fraction of metagenomic data from understudied and diverse habitats such as soil, and even well-studied environments such as human gut can produce significant uncharacterized metagenome content (CITE).

As protein sequence is more conserved than the underlying DNA sequence, protein-based comparisons have long been the gold-standard approach across larger evolutionary distances [@blast; @diamond].
Protein-based metagenomics taxonomic classification approaches show increased sensitivity, both using assembly-based fast-alignment approaches and alignment-free analysis using 6-frame translation of nucleotide sequence [@breit_classif; @mmseqs_tax; @catbat; @meganlr; @kaiju; @mash_screen].
However, for fast and scalable alignment-free methods, increased classification sensitivity is often accompanied by a concomitant decrease in specificity, with some methods reporting results with false positive rates as high as 25% on short read metagenomic datasets prior to thresholding [@portik_tax].
As databases continue to increase in size, there remains a need for fast and accurate alignment-free taxonomic classification with protein-level sensitivity.

Here, we demonstrate the utility of protein k-mer comparisons for both assembly-based and assembly-free metagenomic analyses, including Average Amino Acid Identity (AAI) estimation and taxonomic classification.
We use FracMinhash subsampling to facilitate conducting these comparisons at scale and increase accuracy relative to MinHash methods [@sourmash_gather].

FracMinHash is a MinHash variant for selecting and hashing a set of representative k-mers from a sequence dataset [@sourmash_gather]. Unlike traditional MinHash, FracMinHash sketches scale with the size of the dataset, meaning each sketch is comprised of the chosen fraction of k-mers in the input dataset, rather than a chosen number of k-mers.
Downsampling sequencing datasets in this way enables estimation of the Containment Index, which has been shown to facilitate genome discovery and permit more accurate estimation of genomic distance, particularly for genomes of very different lengths [@koslicki_containment; @dashing; @mash_screen; @metalign].

Pairwise Overall Genome Relatedness indices (OGRI's) such as ANI and its protein counterpart, Average Amino Acid Identity (AAI) have shown lasting utility for genome relatedness and phylogenomic analysis.
Traditional ANI and AAI describe the sequence similarity of all orthologous genes, either in nucleotide or protein space, respectively.
Both have been shown to be robust measures of overall pairwise genome relatedness even for highly incomplete datasets, such as those comprised of only ~4% of the genome or 100 genes [@doi:10.1128/AEM.01398-06; @doi:10.1038/ismej.2017.113].
While traditional alignment-based estimation of ANI is computationally intensive, sketching-based estimates and sketching-facilitated estimates have permitted ANI calculations at the scale of whole-databases [@doi:10.1186/s13059-016-0997-x; @doi:10.1186/s13059-019-1841-x; @doi:10.1038/s41467-018-07641-9].
Containment estimates have also been shown to correlate with Mash Distance, a proxy for Average Nucleotide Identity (ANI) [@mash_screen].
Recently, Hera et. al (2022) [@dk_fracminhash] introduced ANI estimation from FracMinHash containment estimates, while accounting for the non-independence of mutated k-mers [@blanca]. 
Here, we apply the same approach to protein k-mer containment to generate alignment-free AAI estimates.

Here, we show that protein k-mer containment sketches can be used to find similarity across large evolutionary distances and robustly estimate Average Amino Acid Identity.
We demonstrate that FracMinHash Containment estimates can robustly estimate Average Amino Acid Identity across a range of evolutionary distances.
We then use FracMinHash comparison methods to tackle two classification challenges: taxonomic classification of assembled genomes, and compositional analysis of metagenomes.
Taken together, these results suggest that protein FracMinHash analyses can be used for metagenome sequence analysis, and may be particularly useful when sequenced organisms are more distantly related to organisms available in reference databases.


<!---
**Taxonomic Classification -- current approaches + issues**
Using these protein FracMinHashes with protein containment searches, we can conduct taxonomic classification on both assembled genomes and raw metagenome samples.
Recently, [@sourmash_gather] leveraged containment analysis with a "minimum metagenome cover" approach to find the best set of reference genomes that cover the metagenome sequence.
Here, we use this minimum set cover approach with protein FracMinHash to directly classify raw metagenome reads and to classify assembled genomes/proteomes.
..discuss min-set-cov /gather

BLASTX-style classification -- proteins increase sensitivity.
- Kaiju[@doi:10.1038/ncomms11257] uses protein-level metagenomic classification using 6-frame translation, though Kaiju uses a seed-extend approach to generate protein alignments (minimum match length 11aa), rather than a purely k-mer based approach.
- sensitivity/specificity tradeoff -- longer matches (minimum match length 11 instead of 12 for MEM yielded larger gains in sensitivity for a relatively small(er) loss in precision).
- MEGAN-Prot (uses DIAMOND) [@doi:10.1002/cpz1.59] uses a BLASTX style approach
- review incl. protein methods [@doi:10.1093/bib/bbx120]

When conducting 
In many cases, researchers don't need an 

While many methods provide a annotations linked to individual sequencing reads, in many cases, researchers are 
#### Notes
While ANI has emerged as the most widely-accepted method for delimiting species boundaries, AAI thresholds have been proposed for distinguishing higher taxonomic ranks [@genomic_std; @doi:10.1016/j.mib.2007.08.006; @doi:10.1038/ismej.2017.113].
- AAI::phylogeny https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1236649/
Unused
Alignment-based estimates can fail at larger evolutionary distances.
 rRNA amplicon surveys may underestimate bacterial diversity [@doi:10.1128/AEM.00014-18].

 , using genomes within the GTDB taxonomy as benchmarking test set.

We extend FracMinHash distance estimation to protein k-mers and demonstrate distance estimation across related genomes using the GTDB taxonomy.
FracMinHash containment estimates work well for genome pairs of varying lengths and for compositional analysis of metagenome samples.

FracMinHash sketches can be used for robust assembly and alignment-free pairwise relatedness estimation that can be used on sequences separated by a wide range of evolutionary distances.

Here, we demonstrate that the utility of FracMinHash protein containment, both used directly and a an approximation of ANI and AAI, for taxonomic classification and phylogenomic reconstruction for species across the tree of life.

--->

