## Results



K-mer analysis methods enable similarity detection as low as a single shared k-mer between divergent genomes.
As a result, exact matching long nucleotide k-mers can be used for taxonomic classification between closely related genomes, including at the strain, species, and genus level (k-mer lengths 51, 31, and 21, respectively).
At larger evolutionary distances, accumulated nucleotide divergence limits the utility of exact nucleotide k-mer matching.
As protein sequences are more conserved than their coding nucleotide sequences, exact matching of protein k-mers can enable similarity analysis across larger evolutionary distances.
To start, we choose protein k-mer sizes that correspond to known informative DNA ksizes, 21 (protein k-mer length: 7) and 31 (closest protein k-mer length: 10).

<!---
NOTE: worth looking at k=17 /51???.
--->

**genomes in same genus, fraction of k-mers in common at each ksize? ALL THE K-MERS**


### GTDB databases

The Genome Taxonomy Database (GTDB) provides a genome-based taxonomy for bacterial and archaeal genomes [@doi:10.1038/s41587-020-0501-8].
The most recent verion, rs202, contains 258,407 genomes from within 47,895 species.
This database provides a wide array of genomes for assessing the utility of protein k-mers for bacterial and archaeal similarity estimation and taxonomic classification.

To make analyses at this scale tractable, we downsample k-mers using `sourmash` Scaled MinHash sketches, with a scaling factor of 1000 for nucleotide k-mers (keep ~1/1000 k-mers) and 100 for protein k-mers (keep ~1/100 protein k-mers).
For most genomes, both genomic and protein fastas were available for download from NCBI. In remaining cases (n=36,632), genome fastas were translated into protein sequence via PRODIGAL (CITE) prior to sketching. 

We built `sourmash` databases for GTDB, `rs202` version, both for the species-representatives subset, and for all genomes.
These are available as part of the `Prepared Databases` section of the `sourmash` documentation, and archived on OSF [https://osf.io/t3fqa/] /Zenodo???.

### Different k-mer lengths provide resolution at different taxonomic ranks

`sourmash lca rankinfo`.

For each database and k-mer type, we first assessed the number of k-mers specific to each taxonomic rank: k-mers specific to a geonome were only present in one genome in the entire database, k-mers specific to a species were found in at least two genomes of the same species, etc. K-mers specific to a "superkingdom" were found in genomes spanning at least two phyla. While these characterizations are greatly impacted by the genomes included in the database, the differences observed between k-mer types and sizes suggests that different k-mer sizes may provide resolution at different taxonomic ranks. 

<!---
note 31, 51 --- maybe partially a result of database issues, e.g. not all species have multiple members; sometimes all members are closely related.
--->

For all DNA k-mer sizes, the majority of k-mers are present in only a single species, with only a few k-mers shared across genera.
Long nucleotide k-mers have already been shown to be useful for comparing genomes within the same genus or species.
Only at a dna k-mer size of 21 are a significant fraction of k-mers present in genomes shared across different families or even phyla.
In contrast, all protein k-mer sizes contain a portion of k-mers that are shared across genera and above.
At a protein k-mer size of 7, over 80% of k-mers are present in genomes found in more than one phylum, while at a protein k-size of 10, the number of genome-specific k-mers is more similar to that observed for nucleotide k-mers. 


![**Fraction of k-mers specific to taxonomic rank**
For the GTDB-RS202 database, the majority of nucleotide k-mers are specific to (unique at) a specific genome, species, or genus. Few k-mers are shared across superkingdoms, though these do exist at k=21. K-mers shared at such a high level are indicative of high k-mer homoplasy: the presence of k-mers that are identical by chance rather than evolutionary descent. ](images/gtdb-rs202.lca_f_aggregated_kmers.png){#fig:gtdb-kmers height=2in}

GTDB rankinfo: 

  - xx% of DNA k-mers (k=21) are shared within-species
  - yy% of protein k-mers are shared within-species
  - zz% of DNA k-mers are shared within-genus ... etc 


Percent/Number of shared k-mers between members of same species/genus?



### Protein k-mers enable alignment-free comparisons at increased evolutionary distances



As the number of genomes per taxon varies widely across GTDB, comparisons across the entire GTDB database may be impacted by the particular genoems included in the database. To minimize any database biases, we developed a randomized subset of the GTDB database to assess the utility of protein k-mers across a wide range of evolutionary distances.
For each genus with at least two species clusters in GTDB, one representative genome was randomly selected as an "anchor" genome.
Then, one additional genome was selected from the GTDB representative genomes matching the anchor's taxonomy at each higher taxonomic rank.
This "evolutionary path" consists of seven genomes: an anchor genome, a genome matching anchor taxonomy down to the genus level, one matching anchor taxonomy to the family level, one matching to the order level, and so on.
This creates a gradient of similarity, where comparisons to the anchor genome range from genus-level to superkingdom-level.
_OR SPECIES VERSION:
This "evolutionary path" consists of eight genomes: an anchor genome, a genome matching anchor taxonomy down to the species level, one matching anchor taxonomy to the genus level, one matching to the family level, and so on.
This creates a gradient of similarity, where comparisons to the anchor genome range from species-level to superkingdom-level._

Path selection using the representative genomes in **GTDB release 95** resulted in 2957 paths comprised of 6690 unique genomes (6543 Bacteria, 237 Archaea).
These paths include genome comparisons across 33 phyla (29 Bacteria, 4 Archaea), covering roughly a quarter of the 129 phyla (111 Bacteria, 18 Archaea) in GTDB release 95.
While paths are limited to taxonomies with at least two GTDB representative genomes for each taxonomic rank, these paths provide a rich resource for comparisons at increasing evolutionary distances. 

Path selection using the representative genomes in GTDB rs202 resulted in 4095 paths comprised of 9213 unique genomes (8790 Bacteria, 333 Archaea).
_These paths include genome comparisons across 33 phyla (29 Bacteria, 4 Archaea), covering roughly a quarter of the 129 phyla (111 Bacteria, 18 Archaea) in GTDB release 95._
While paths are limited to taxonomies with at least two GTDB representative genomes for each taxonomic rank, these paths provide a rich resource for comparisons at increasing evolutionary distances. 

**_to do: combine these two figures by just comparing species-level for each gtdb representative used in 2nd figure_**
![**More protein k-mers are shared at genus level** CAPTION](images/pseudomonas_jaccard_vs_containment_prot10.png)

![**Protein k-mers are shared at higher taxonomic ranks** CAPTION](images/anchor-containment.nucl-prot.png)





<!--compare heatmap w/ max containment for subset of gtdb data?-->
<!---
For , e.g. Pseudomonas, XX% of k-mers are 	shared within the chosen/published genomes within species. For all published genomes within the genus, a median of xx% of k-mers are shared between genomes of one species and genomes of the a different species in the same genus.


== median or mean containment at rank?
containent = % of a genome's k-mers that are shared
-- do using ALL of gtdb, BUT, start with just a single set of genomes.. e.g. Pseudomonas? == similar to "shared k-mers" paper [@doi:10.24072/pci.genomics.100001]

![**Protein k-mer containment facilitates genus-level comparisons**
10k pseudomonas genome sequences, median containment at each alphabet](images/pseudomonas_jaccard_vs_containment_prot10.png){#fig:evolpathsContain}
--->


### Accurate distance estimation from k-mer containment


**__to do: A. redo with MRCC estimate; use only best graph (or maybe two ksizes)__**
![**Scaled MinHash AAI vs CompareM**
GTDB Evolpaths dataset](images/gtdb95-evolpaths.AAI-concordance.png){#fig:evolpathsAAIvsCompareM height=2in}

### **Anchor containment** enables comparisons directly from DNA sequence

** figure: AAI from translated nucleotide --> reference protein**


### Robust Taxonomic classification from protein k-mer containment



### Metagenome breakdown using protein k-mers

*** use a mock metagenome, then a evolutionarily distant metagenome. compare the % of genome recovred with DNA, protein at diff ksizes. Genome grist it, basically.


THIS USES BOTH THE 6-FRAME translation and sourmash tax!!! containment improtant, etc. I think this is the way to go.



### Phylogenetic Reconstruction from k-mer Amino Acid identity
NO, just leave this out

![**K-mer Based Sequence Identity by Lowest Common Taxon**
GTDB Evolpaths dataset](images/anchor-mcANI-AAI.boxen.protnucl.png){#fig:evolpathsANIAAI}










### Lost Bits


Long dna k-mers ~~ short protein k-mers

while shorter dna k-mers might be shared across more sequence, you increase the risk for result in"shared, non-homologous k-mers" (k-mer homoplasy). A protein k-mer of length `10` coverse 30 base pairs in nuof nucleotide sequence 

DR